File: src/main/java/com/vc/hard/L51.java
package com.vc.hard;

import java.util.*;

class L51 {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<List<String>>();
        char[][] state = new char[n][n];
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                state[i][j] = '.';
            }
        }
        solve(0, n, state, res);
        return res;
    }

    private void solve(int row, int n, char[][] state, List<List<String>> res) {
        //print(state);
        if(row == n) {
            List<String> list = new ArrayList<String>();
            for(int i = 0; i < state.length; i++) {
                StringBuilder str = new StringBuilder();
                for(int j = 0; j < state[i].length; j++) {
                    str.append(state[i][j]);
                }
                list.add(str.toString());
                str.setLength(0);
            }
            res.add(list);
        }
        else {
            for(int col = 0; col < n; col++) {
                if(isValid(row, col, n, state)) {
                    state[row][col] = 'Q';
                    solve(row + 1, n, state, res);
                    state[row][col] = '.';
                }
            }
        }
    }

    private void print(char[][] state) {
        System.out.println("==========================================");
        for(int i = 0; i < state.length; i++) {
            for(int j = 0; j < state[0].length; j++) {
                System.out.print(state[i][j]+" ");
            }
            System.out.println("");
        }
        System.out.println("==========================================");
    }

    private boolean isValid(int row, int col, int n, char[][] state) {
        int drow = row;
        int dcol = col;
        int dacol = col;
        //System.out.println("Trying row: "+row+" col: "+col);
        for(int i = 0; i < n; i++) {
            //Checking for same column
            if(state[i][col] == 'Q') {
                //System.out.println("Failed for column validation in row: "+i+" and col: "+col);
                return false;
            }

            //Checking diagonal
            if(drow >= 0 && dcol >= 0 && state[drow][dcol] == 'Q') {
                //System.out.println("Failed for diagonal validation in row: "+drow+" and col: "+dcol);
                return false;
            }

            //Checking anti-diagonal
            if(drow >= 0 && dacol < n && state[drow][dacol] == 'Q') {
                //System.out.println("Failed for Anti-diagonal validation in row: "+drow+" and col: "+dacol);
                return false;
            }

            drow--;
            dacol++;
            dcol--;
        }
        return true;
    }
}

File: src/main/java/com/vc/hard/L51.scala
package com.vc.hard

object L51 {
  def solveNQueens(n: Int): List[List[String]] = {
    import scala.collection.mutable
    val res = new mutable.ListBuffer[List[String]]()
    val arr = Array.ofDim[Char](n, n)
    arr.indices.foreach(i => {
      arr(i).indices.foreach(j => {
        arr(i)(j) = '.'
      })
    })

    def solve(row: Int): Unit = {
      if(row == n) {
        val l = new mutable.ListBuffer[String]()
        arr.indices.foreach(i => {
          l += arr(i).mkString("")
        })
        res += l.toList
      }
      else {
        (0 until n).foreach(col => {
          arr(row)(col) = 'Q'
          if(isValid(row, col)) solve(row + 1)
          arr(row)(col) = '.'
        })
      }
    }

    def isValid(row:Int, col: Int): Boolean = {
      var i = 0
      var j = col
      while(i < row) {
        if(arr(i)(j) == 'Q') return false
        i += 1
      }

      i = row - 1
      j = col - 1
      while(i >= 0 && j >= 0) {
        if(arr(i)(j) == 'Q') return false
        i -= 1
        j -= 1
      }

      i = row - 1
      j = col + 1
      while(i >= 0 && j < n) {
        if(arr(i)(j) == 'Q') return false
        i -= 1
        j += 1
      }
      true
    }
    solve(0)
    res.toList
  }
}
